### ISSUE TITLE
neural_sde example and test_brownian.py throw ValueError: ShapeDtypeStruct: dtype must be specified.

### ISSUE THREAD

**Issue by bn2302** [patrick-kidger/diffrax#225]
When running the neural_sde jupyternotebook, I receive the following error message after calling `main()` cell:

```python

No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[9], line 1
----> 1 main()

Cell In[8], line 28, in main(initial_noise_size, noise_size, hidden_size, width_size, depth, generator_lr, discriminator_lr, batch_size, steps, steps_per_print, dataset_size, seed)
     17 (
     18     data_key,
     19     generator_key,
   (...)
     24     sample_key,
     25 ) = jrandom.split(key, 7)
     26 data_key = jrandom.split(data_key, dataset_size)
---> 28 ts, ys = get_data(data_key)
     29 _, _, data_size = ys.shape
     31 generator = NeuralSDE(
     32     data_size,
     33     initial_noise_size,
   (...)
     38     key=generator_key,
     39 )

    [... skipping hidden 15 frame]

Cell In[6], line 20, in get_data(key)
     17 def diffusion(t, y, args):
     18     return 2 * sigma * t / t1
---> 20 bm = diffrax.UnsafeBrownianPath(shape=(), key=bm_key)
     21 drift = diffrax.ODETerm(drift)
     22 diffusion = diffrax.ControlTerm(diffusion, bm)

File ~/github/diffrax/.conda/lib/python3.10/site-packages/equinox/module.py:94, in _ModuleMeta.__call__(cls, *args, **kwargs)
     92 object.__setattr__(self, "__class__", initable_cls)
     93 try:
---> 94     cls.__init__(self, *args, **kwargs)
     95 finally:
     96     object.__setattr__(self, "__class__", cls)

File ~/github/diffrax/.conda/lib/python3.10/site-packages/diffrax/brownian/path.py:46, in UnsafeBrownianPath.__init__(self, shape, key)
     40 def __init__(
     41     self,
     42     shape: Union[Tuple[int, ...], PyTree[jax.ShapeDtypeStruct]],
     43     key: "jax.random.PRNGKey",
     44 ):
     45     self.shape = (
---> 46         jax.ShapeDtypeStruct(shape, None) if is_tuple_of_ints(shape) else shape
     47     )
     48     self.key = key
     49     if any(
     50         not jnp.issubdtype(x.dtype, jnp.inexact)
     51         for x in jtu.tree_leaves(self.shape)
     52     ):

File ~/github/diffrax/.conda/lib/python3.10/site-packages/jax/_src/api.py:3210, in ShapeDtypeStruct.__init__(self, shape, dtype, named_shape, sharding)
   3208 self.shape = tuple(shape)
   3209 if dtype is None:
-> 3210   raise ValueError("ShapeDtypeStruct: dtype must be specified.")
   3211 self.dtype = dtype if core.is_opaque_dtype(dtype) else np.dtype(dtype)
   3212 if sharding is not None:

ValueError: ShapeDtypeStruct: dtype must be specified.

```

`test_brownian.py` also throws the same error.

```python
============================= test session starts ==============================
platform linux -- Python 3.10.9, pytest-7.2.1, pluggy-1.0.0
rootdir: /home/user/github/diffrax
plugins: typeguard-2.13.3, jaxtyping-0.2.11
collected 2 items

test/test_brownian.py FF                                                 [100%]

=================================== FAILURES ===================================
___________________ test_shape_and_dtype[UnsafeBrownianPath] ___________________

ctr = <class 'diffrax.brownian.path.UnsafeBrownianPath'>
getkey = <function getkey.<locals>._getkey at 0x7f34382bd510>

    @pytest.mark.parametrize(
        "ctr", [diffrax.UnsafeBrownianPath, diffrax.VirtualBrownianTree]
    )
    def test_shape_and_dtype(ctr, getkey):
        t0 = 0
        t1 = 2
    
        shapes = (
            (),
            (0,),
            (1, 0),
            (2,),
            (3, 4),
            (1, 2, 3, 4),
            {
                "a": (1,),
                "b": (2, 3),
            },
            (
                (1, 2),
                (
                    (3, 4),
                    (5, 6),
                ),
            ),
        )
    
        dtypes = (
            None,
            None,
            None,
            jnp.float16,
            jnp.float32,
            jnp.float64,
            {"a": None, "b": jnp.float64},
            (jnp.float16, (jnp.float32, jnp.float64)),
        )
    
        def is_tuple_of_ints(obj):
            return isinstance(obj, tuple) and all(isinstance(x, int) for x in obj)
    
        for shape, dtype in zip(shapes, dtypes):
            # Shape to pass as input
            if dtype is not None:
                shape = jtu.tree_map(
                    jax.ShapeDtypeStruct, shape, dtype, is_leaf=is_tuple_of_ints
                )
    
            if ctr is diffrax.UnsafeBrownianPath:
                path = ctr(shape, getkey())
                assert path.t0 is None
                assert path.t1 is None
            elif ctr is diffrax.VirtualBrownianTree:
                tol = 2**-5
                path = ctr(t0, t1, tol, shape, getkey())
                assert path.t0 == 0
                assert path.t1 == 2
            else:
                assert False
    
            # Expected output shape
            if dtype is None:
>               shape = jtu.tree_map(
                    jax.ShapeDtypeStruct, shape, dtype, is_leaf=is_tuple_of_ints
                )

test/test_brownian.py:82: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.conda/lib/python3.10/site-packages/jax/_src/tree_util.py:207: in tree_map
    return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))
.conda/lib/python3.10/site-packages/jax/_src/tree_util.py:207: in <genexpr>
    return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <[AttributeError("'ShapeDtypeStruct' object has no attribute 'named_shape'") raised in repr()] ShapeDtypeStruct object at 0x7f343831fe80>
shape = (), dtype = None, named_shape = None, sharding = None

    def __init__(self, shape, dtype, named_shape=None, sharding=None):
      self.shape = tuple(shape)
      if dtype is None:
>       raise ValueError("ShapeDtypeStruct: dtype must be specified.")
E       ValueError: ShapeDtypeStruct: dtype must be specified.

.conda/lib/python3.10/site-packages/jax/_src/api.py:3210: ValueError
------------------------------ Captured log call -------------------------------
WARNING  jax._src.lib.xla_bridge:xla_bridge.py:362 No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
__________________ test_shape_and_dtype[VirtualBrownianTree] ___________________

ctr = <class 'diffrax.brownian.tree.VirtualBrownianTree'>
getkey = <function getkey.<locals>._getkey at 0x7f34382ed990>

    @pytest.mark.parametrize(
        "ctr", [diffrax.UnsafeBrownianPath, diffrax.VirtualBrownianTree]
    )
    def test_shape_and_dtype(ctr, getkey):
        t0 = 0
        t1 = 2
    
        shapes = (
            (),
            (0,),
            (1, 0),
            (2,),
            (3, 4),
            (1, 2, 3, 4),
            {
                "a": (1,),
                "b": (2, 3),
            },
            (
                (1, 2),
                (
                    (3, 4),
                    (5, 6),
                ),
            ),
        )
    
        dtypes = (
            None,
            None,
            None,
            jnp.float16,
            jnp.float32,
            jnp.float64,
            {"a": None, "b": jnp.float64},
            (jnp.float16, (jnp.float32, jnp.float64)),
        )
    
        def is_tuple_of_ints(obj):
            return isinstance(obj, tuple) and all(isinstance(x, int) for x in obj)
    
        for shape, dtype in zip(shapes, dtypes):
            # Shape to pass as input
            if dtype is not None:
                shape = jtu.tree_map(
                    jax.ShapeDtypeStruct, shape, dtype, is_leaf=is_tuple_of_ints
                )
    
            if ctr is diffrax.UnsafeBrownianPath:
                path = ctr(shape, getkey())
                assert path.t0 is None
                assert path.t1 is None
            elif ctr is diffrax.VirtualBrownianTree:
                tol = 2**-5
                path = ctr(t0, t1, tol, shape, getkey())
                assert path.t0 == 0
                assert path.t1 == 2
            else:
                assert False
    
            # Expected output shape
            if dtype is None:
>               shape = jtu.tree_map(
                    jax.ShapeDtypeStruct, shape, dtype, is_leaf=is_tuple_of_ints
                )

test/test_brownian.py:82: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.conda/lib/python3.10/site-packages/jax/_src/tree_util.py:207: in tree_map
    return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))
.conda/lib/python3.10/site-packages/jax/_src/tree_util.py:207: in <genexpr>
    return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <[AttributeError("'ShapeDtypeStruct' object has no attribute 'named_shape'") raised in repr()] ShapeDtypeStruct object at 0x7f34381b2940>
shape = (), dtype = None, named_shape = None, sharding = None

    def __init__(self, shape, dtype, named_shape=None, sharding=None):
      self.shape = tuple(shape)
      if dtype is None:
>       raise ValueError("ShapeDtypeStruct: dtype must be specified.")
E       ValueError: ShapeDtypeStruct: dtype must be specified.

.conda/lib/python3.10/site-packages/jax/_src/api.py:3210: ValueError
------------------------------ Captured log call -------------------------------
WARNING  jax._src.lib.xla_bridge:xla_bridge.py:362 No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
-------------- generated xml file: /tmp/tmp-5396myg7L9xsntGm.xml ---------------
=========================== short test summary info ============================
FAILED test/test_brownian.py::test_shape_and_dtype[UnsafeBrownianPath] - Valu...
FAILED test/test_brownian.py::test_shape_and_dtype[VirtualBrownianTree] - Val...
============================== 2 failed in 1.44s ===============================
ERROR conda.cli.main_run:execute(47): `conda run python /home/user/.vscode-server/extensions/ms-python.python-2023.2.0/pythonFiles/get_output_via_markers.py -m pytest --rootdir /home/user/github/diffrax --override-ini junit_family=xunit1 --junit-xml=/tmp/tmp-5396myg7L9xsntGm.xml ./test/test_brownian.py::test_shape_and_dtype` failed. (See above for error)
```

Is there a version of jax, which I can try to pinpoint the cause of this behavior, or any settings I can change?

Environment:

```
Conda - Python 3.10 - Linux x64 (WSL)

Package                  Version
------------------------ -----------
absl-py                  1.4.0
anyio                    3.6.2
argon2-cffi              21.3.0
argon2-cffi-bindings     21.2.0
arrow                    1.2.3
asttokens                2.2.1
attrs                    22.2.0
backcall                 0.2.0
beautifulsoup4           4.11.2
bleach                   6.0.0
certifi                  2022.12.7
cffi                     1.15.1
chex                     0.1.6
comm                     0.1.2
contourpy                1.0.7
cycler                   0.11.0
debugpy                  1.6.6
decorator                5.1.1
defusedxml               0.7.1
diffrax                  0.2.2
dm-tree                  0.1.8
equinox                  0.9.2
exceptiongroup           1.1.0
executing                1.2.0
fastjsonschema           2.16.2
fonttools                4.38.0
fqdn                     1.5.1
idna                     3.4
iniconfig                2.0.0
ipykernel                6.21.1
ipython                  8.10.0
ipython-genutils         0.2.0
ipywidgets               8.0.4
isoduration              20.11.0
jax                      0.4.3
jaxlib                   0.4.3
jaxtyping                0.2.11
jedi                     0.18.2
Jinja2                   3.1.2
jsonpointer              2.3
jsonschema               4.17.3
jupyter                  1.0.0
jupyter_client           8.0.2
jupyter-console          6.5.0
jupyter_core             5.2.0
jupyter-events           0.6.3
jupyter_server           2.2.1
jupyter_server_terminals 0.4.4
jupyterlab-pygments      0.2.2
jupyterlab-widgets       3.0.5
kiwisolver               1.4.4
MarkupSafe               2.1.2
matplotlib               3.6.3
matplotlib-inline        0.1.6
mistune                  2.0.5
nbclassic                0.5.1
nbclient                 0.7.2
nbconvert                7.2.9
nbformat                 5.7.3
nest-asyncio             1.5.6
notebook                 6.5.2
notebook_shim            0.2.2
numpy                    1.24.2
opt-einsum               3.3.0
optax                    0.1.4
packaging                23.0
pandocfilters            1.5.0
parso                    0.8.3
pexpect                  4.8.0
pickleshare              0.7.5
Pillow                   9.4.0
pip                      22.3.1
platformdirs             3.0.0
pluggy                   1.0.0
prometheus-client        0.16.0
prompt-toolkit           3.0.36
psutil                   5.9.4
ptyprocess               0.7.0
pure-eval                0.2.2
pycparser                2.21
Pygments                 2.14.0
pyparsing                3.0.9
pyrsistent               0.19.3
pytest                   7.2.1
python-dateutil          2.8.2
python-json-logger       2.0.4
PyYAML                   6.0
pyzmq                    25.0.0
qtconsole                5.4.0
QtPy                     2.3.0
rfc3339-validator        0.1.4
rfc3986-validator        0.1.1
scipy                    1.10.0
Send2Trash               1.8.0
setuptools               65.6.3
six                      1.16.0
sniffio                  1.3.0
soupsieve                2.3.2.post1
stack-data               0.6.2
terminado                0.17.1
tinycss2                 1.2.1
tomli                    2.0.1
toolz                    0.12.0
tornado                  6.2
traitlets                5.9.0
typeguard                2.13.3
typing_extensions        4.4.0
uri-template             1.2.0
wcwidth                  0.2.6
webcolors                1.12
webencodings             0.5.1
websocket-client         1.5.1
wheel                    0.37.1
widgetsnbextension       4.0.5
```

**Comment by patrick-kidger** [comment]
Thanks for the heads-up! This is due to a change in JAX's behaviour (https://github.com/google/jax/pull/13814). This should be fixed in the next release of Diffrax. (In the next few days.)

**Comment by mdiamore** [comment]
I was playing around with the SDE example  as well,  I managed to get it working by changing  

`jax.ShapeDtypeStruct(shape, jax.dtypes.canonicalize_dtype(None))`  

to 

 `jax.ShapeDtypeStruct(shape, jax.dtypes.result_type(None))`

in `/diffrax/brownian/paths.py` and `/diffrax/brownian/tree.py`

However it does break some of the tests and I'm not sure what any other downstream implications are but the example does complete

**Comment by patrick-kidger** [comment]
This should now have been fixed in v0.3.0.
